### ISSUE TITLE
💡[Need Feedback] - Native Typescript Support Strategy in core library

### ISSUE THREAD

**Issue by cliffordfajardo** [react-grid-layout/react-grid-layout#1683]
## Describe the Feature Request

**As user of `react-grid-layout`:**
1. I would like types to be colocated close to the actual source code
   - That is, have the types from `react-grid-layout` defined in this same repo as opposed to installing a 3rd party dependency `@types/react-grid-layout`
   - In the long run, this reduces operational burden from maintainers and members of the community who would like to improve the types. Contributing to or enhancing the types will no longer require forking and maintaining 2 packages, waiting for 1 to get merged, having things be out of sync etc. The easier the process is, the easier it is to contribute; this would help minimize friction points for contributing 
   
**As a potential contributor of `react-grid-layout` codebase:**
- having the types colocated with the code will allow me more quickly familiarize myself with the codebase
- i no longer need to refer to 2 repositories `@types/react-grid-layout` for verifying types etc
- over the long run:
   - reduces bugs due to type safety
   - increases productivity & saves us time since IDE/editor's have more information available
- reduces friction points for contributing

<br/>
<br/>

### Proposed Initial Implementation Strategy
**The main goal is to have the code working as before with very minimal friction**
- consumers of the library will not notice any new changes & this will not require changes to `@types/react-grid-layout`
- the code will be compiled in loose mode, expect to see lots of `any`'s in complex areas of the codebase

**0. create a new branch for typescript.**

**1. Rename `.js` files to `.ts / tsx`**
This script below demonstrates how relatively straightforward converting all JS files to typescript can be. This will handle renaming the file with `git mv` so git doesn't track this file as a deletion or entirely new file.
```
# rename all JSX files in src/ to TSX (reactjs example)
find src -name '*.jsx' -exec bash -c 'git mv "$0" "${0%.jsx}.tsx"' "{}" \;

# rename all JS files in src/ to TS
find src -name '*.js' -exec bash -c 'git mv "$0" "${0%.js}.ts"' "{}" \;

# rename all JSX files in tests/ to TSX
find tests -name '*.jsx' -exec bash -c 'git mv "$0" "${0%.jsx}.tsx"' "{}" \;

# rename all JSX files in tests/__snapshots/ to TSX
find tests -name '*.jsx.snap' -exec bash -c 'git mv "$0" "${0%.jsx.snap}.tsx.snap"' "{}" \;

# rename all JS files in tests/ to TS
find tests -name '*.js' -exec bash -c 'git mv "$0" "${0%.js}.ts"' "{}" \;
```

2. we will likely need to update some of the build system's `devDependency` to support typescript
- update test suite to support reading typescript/JSX files
- update webpack to support reading typescript files

3. test that the UI works
4. ensure all tests pass as before 

### Other Considerations
- There are other codemod tools that exist (ex: [airbnb/ts-migrate](https://github.com/airbnb/ts-migrate)), but they tend to be more complicated and require a certain folder layout, whereas this bash script is simple and is project agnostic

**Comment by cliffordfajardo** [comment]
@STRML - what do you think? I didn't want to start writing any code without input from the community and/or waste folk's time

**Comment by STRML** [comment]
Hi. I think converting the repo to TS is a good idea, and would support it. However, there are two things you may be underestimating:

1. Keeping Flow types available, either on https://github.com/flow-typed/flow-typed, by converting automatically from TS definitions, or by hand
2. Converting from Flow to TS will take some manual work aside from just renaming the files.

That said, if you're wanting to tackle that, I would absolutely support it and spend time reviewing.

**Comment by linkb15** [comment]
https://transform.tools/flow-to-typescript-declaration

@cliffordfajardo @STRML this could be simpler with tools like this to help you guys out converting it to type declaration.

**Comment by marcus-pousette** [comment]
Spent the day porting this lib in a [fork](https://github.com/dao-xyz/react-grid-layout-next).
- TypeScript 
- Functional React
- Lerna monorepo, where one package is the lib, other is examples (with vite build rather than webpack)
- React 18
- Removal of all prop-types things. 

Tests are not working at the moment, but the code is fully functional (by playing around with all examples).

Could perhaps do this as a PR, to back-propagate changes, else it could perhaps help with some work.

**Comment by STRML** [comment]
That sounds great. I'd rather it be a TS lib given Flow has lost the battle.

**Comment by timoisalive** [comment]
Hey! Is this a closed ticket or can we perhaps expect a TS version sometime? I just figured `@types/react-grid-layout` isn't 100% accurate. At least the `resizeHandle` type seems wrong...

**Comment by marcus-pousette** [comment]
> Hey! Is this a closed ticket or can we perhaps expect a TS version sometime? I just figured `@types/react-grid-layout` isn't 100% accurate. At least the `resizeHandle` type seems wrong...

See my comment, where I present my work porting this lib into a functional Typescript project. 

Regarding the issue I think it was automatically closed by being stale

**Comment by timoisalive** [comment]
> @timoisalive See my comment, where I present my work porting this lib into a functional Typescript project.

@marcus-pousette That's awesome, I'm just about to check it out!

It would however be nice if your work could be merged, so we could stay on the main library instead of a fork, so keeping my fingers crossed...

Thanks for your effort though! 👏

**Comment by marcus-pousette** [comment]
Thanks. Yeap I hope/wish it can be merged back to!

There are no real breaking changes except that almost every single file in the project has been changed. During the migration I did not take pay a lot of attention of making the test suite work, but to fix critical bugs that I had for my end use-case. If this fork will be merged back there are two things needed

- Rewriting the test suite so it is up to date (and also improved testing more edge cases that were previously missed). 
- Bump major version
